### PR TITLE
REF: let EAs override WrappedCythonOp groupby implementations

### LINKED ISSUE #43682: ENH: generic implementation of ea_wrap_cython_operation
Create a property function to decide what to do for EAs in `_ea_wrap_cython_operation` for groupby ops.
xref : https://github.com/pandas-dev/pandas/pull/43634#discussion_r712140219

https://github.com/pandas-dev/pandas/blob/439acc99b2968ec8062eee05af7346acea20915b/pandas/core/groupby/ops.py#L318-L333

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #43682

Goal is to allow EAs (including our own pyarrow-backed) to implement performant GroupBy reductions/transforms without necessarily having to convert to numpy.

Analogous to #51003 (quantile) and #51116 (any, all, std).

This implements EA.grouby_op which takes a "how" keyword to specify which reduction/transform is being done.  By contrast, #51003 implements EA.groupby_quantile and #51116 implements EA.groupby_std and EA.groupby_any_all.  If we go this direction (we should), we should choose one naming scheme for all these methods instead of three slightly different ones.  i.e. either a) shove everything into something like `groupby_op` or b) explicitly have `groupby_foo` for each of the relevant methods.

Also will need docs before moving forward.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
changed groupby_op to _groupby_op and added a docstring making explicit that this API is experimental.

Open to formatting suggestions for the kwargs part of the docstring.

**Comment by jbrockmendel** [comment]
> Is there actually a demand from ExtensionArray authors to be able to implement the groupby reductions? (I mean, I can certainly see that it could be useful for certain kind of extension dtypes, but I would still be wary to implement it if it's not really requested)

No, but I think there is a chicken/egg issue in that some libraries (i have in mind cudf, dask, modin) haven't implemented EAs at all and this is a necessary step to making EAs a viable option for them.

> Or is this more as a way to clean up the dispatching in the groupby code for our own EAs? Because then we should discuss if we actually want to make this public of external EAs, or only do it for our internal ones for now (also the pyarrow-backed arrays could be done internally just like we currently special case the masked arrays)

That is definitely a motivator.  This is much nicer abstraction-wise.  We agreed long ago that ideally our internal EAs would be treated symmetrically with 3rd party EAs.

**Comment by jbrockmendel** [comment]
@MarcoGorelli this is the one i mentioned

**Comment by jbrockmendel** [comment]
@jorisvandenbossche feedback in #51471 seems mostly positive.  any thoughts?

**Comment by jbrockmendel** [comment]
@MarcoGorelli gentle ping

**Comment by jreback** [comment]
+1 here. to the extent this cleans up generic pandas code this is quite valuable (not to mention that EA authors can use this).

**Comment by jbrockmendel** [comment]
@mroeschke gentle ping, this is the way forward for e.g. #51996

**Comment by mroeschke** [comment]
I get this solves https://github.com/pandas-dev/pandas/issues/51996 in the short term, but my only reservation here is the dependence on the existing grouping machinery (the determination of `ids`) w.r.t. pyarrow. It would be nice if we could utilize `pyarrow.Table`'s groupby machinery which then leads to using their groupby aggregation machinery too. If we had some notion of "object that manages EA" (https://github.com/pandas-dev/pandas/issues/51471), that could be a pyarrow Table for `ArrowExtensionArrays` where we can dispatch these groupby ops to.

This also applies probably to other dataframe libraries like cuDF/Dask/etc that we hope could utilize the EA interface, but groupby is a "frame-wise" operation where this concern also applies.

**Comment by jbrockmendel** [comment]
Aside from the 3rd-party concerns, our own EA-specific logic fits much better here than it does in groupby.ops.  I've changed this to use a private name to somewhat safeguard against future changes.

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -1728,6 +1728,81 @@ def map(self, mapper, na_action=None):
...
+        np.ndarray or ExtensionArray
+        """
+        from pandas.core.arrays.string_ import StringDtype
+        from pandas.core.groupby.ops import WrappedCythonOp
+
+        kind = WrappedCythonOp.get_kind_from_how(how)
+        op = WrappedCythonOp(how=how, kind=kind, has_dropped_na=has_dropped_na)
+
+        # GH#43682
+        if isinstance(self.dtype, StringDtype):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:1776]: yes

File: pandas/core/arrays/sparse/array.py [hunk 2]
```diff
@@ -1811,6 +1811,21 @@ def _formatter(self, boxed: bool = False):
...
+        self,
+        *,
+        how: str,
+        has_dropped_na: bool,
+        min_count: int,
+        ngroups: int,
+        ids: npt.NDArray[np.intp],
+        **kwargs,
+    ):
+        raise NotImplementedError(f"{self.dtype} dtype not supported")
```
Comment (mroeschke) [inline pandas/core/arrays/sparse/array.py:1827]: Shouldn't the base class raise a `NotImplementedError`? Or is this to keep the same error message?

**Comment by mroeschke** [comment]
Okay I see the benefit for our EAs currently. Just hoping that you are open to a future API where 3rd party EAs (including pyarrow) could also dictate the grouping + op steps together.

**Comment by jbrockmendel** [comment]
> Just hoping that you are open to a future API where 3rd party EAs (including pyarrow) could also dictate the grouping + op steps together

Absolutely open to it.  I'd love to rip out a bunch of our libgroupby code one day.

**Comment by MarcoGorelli** [comment]
> @MarcoGorelli gentle ping

Apologies, was off last week (though still did some minor things). I've made it a priority to look at this next week

**Review by MarcoGorelli — COMMENTED** [review]
No objections, generally looks good, but I'm not familiar enough with this part of the code to say whether this should be the way to go

**Comment by mroeschke** [comment]
I had https://github.com/pandas-dev/pandas/pull/51166/files#r1139482980 but otherwise looked good

File: pandas/core/arrays/sparse/array.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py:1822]: This is just to keep the existing message

Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py:1822]: Looks like this isn't actually needed.  Will remove in a follow-up (or next commit if it turns out to be necessary)

**Review by mroeschke — APPROVED** [review]

**Review by MarcoGorelli — APPROVED** [review]
Should be fine, looks like the only changes since I last reviewed are minor mypy fixups and merges from upstream/main

File: pandas/core/arrays/masked.py [hunk 3]
```diff
@@ -1382,3 +1382,46 @@ def _accumulate(
...
+        ids: npt.NDArray[np.intp],
+        **kwargs,
+    ):
+        from pandas.core.groupby.ops import WrappedCythonOp
+
+        kind = WrappedCythonOp.get_kind_from_how(how)
+        op = WrappedCythonOp(how=how, kind=kind, has_dropped_na=has_dropped_na)
+
+        # libgroupby functions are responsible for NOT altering mask
+        mask = self._mask
```
Comment (phofl) [inline pandas/core/arrays/masked.py:1405]: We should probably add tests covering this at some point. Not sure if we already have them

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @jbrockmendel

**Pull request merged**